File: LeetCode/src/binarytree/BalancedBinaryTree.java
package binarytree;

public class BalancedBinaryTree {

    public boolean isBalanced(TreeNode root) {
        return dfs(root) != -1;
    }

    private int dfs(TreeNode node) {
        if (node == null) {
            return 0;
        }

        int left = dfs(node.left);
        int right = dfs(node.right);
        int gap = Math.abs(left - right);

        if (gap > 1 || left == -1 || right == -1) return -1;

        return Math.max(left,right) + 1;
    }


    public static void main(String[] args) {
        BalancedBinaryTree solution = new BalancedBinaryTree();

        // Test case
        TreeNode root = new TreeNode(3);
        root.left = new TreeNode(9);
        root.right = new TreeNode(20);
        root.right.left = new TreeNode(15);
        root.right.right = new TreeNode(7);

        System.out.println("Is the Binary Tree Balanced? " + solution.isBalanced(root)); // Output: true
    }
}

File: LeetCode_Python/binarytree/BalancedBinaryTree.py
from typing import Optional
from TreeNode import TreeNode


class BalancedBinaryTree:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True

        def height(node: TreeNode) -> int:
            if not node:
                return 0
            left_height = height(node.left)
            right_height = height(node.right)
            if (
                left_height == -1
                or right_height == -1
                or abs(left_height - right_height) > 1
            ):
                return -1
            return max(left_height, right_height) + 1

        return height(root) != -1


# Time Complexity: O(n), where n is the number of nodes in the tree.
# Space Complexity: O(h), where h is the height of the tree, due to the recursion stack.

if __name__ == "__main__":
    solution = BalancedBinaryTree()

    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)
    root.right.right = TreeNode(6)

    print("Expected Output: True")
    print("Actual Output:", solution.isBalanced(root))

    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)
    root.right.right = TreeNode(6)
    root.left.left.left = TreeNode(7)
    root.left.left.right = TreeNode(8)
    root.left.left.left.left = TreeNode(9)
    root.left.left.left.right = TreeNode(10)
    print("Expected Output: False")
    print("Actual Output:", solution.isBalanced(root))
